### PR TITLE
Add method `getFieldSelectionWithAliases` to class `ResolveInfo`

### PR DESCRIPTION
**Pull request by zephyx**
@see https://github.com/webonyx/graphql-php/issues/1072

That was the simplest way to solve this issue with minimal breaking change due to the presence of $this->arrayMergeDeep(...).

If you have a better idea, no problem, I just needed this feature in my application.

NOTE: composer fix / check have updated several files but I have not modified them, so they are not included here.

### CODE REVIEW

**Comment by zephyx** [comment]
I've made a few changes in order to manage the case where we have a field which is not aliased and the same one aliased.
Before we were loosing the args of the first one, not now with those modifications.

**Comment by spawnia** [comment]
Please add tests.

**Comment by zephyx** [comment]
After writing some tests I have seen an edge case and completely refactored the code and moved it somewhere else.
It's late, I will finish the tests tomorrow or sometime this week :)

File: src/Type/Definition/ResolveInfo.php [hunk 1]
```diff
@@ -214,6 +216,126 @@ public function getFieldSelection(int $depth = 0): array
...
+     *          'aliases' => [
+     *              'id' => [
+     *                  [args] => []
+     *              ]
+     *          ]
+     *      ],
+     *      'nested' => [
+     *           'aliases' => [
+     *               'nested' => [
+     *                  ['args'] => [],
```
Comment (spawnia) [inline src/Type/Definition/ResolveInfo.php:255]: This does not look like a valid array, what is going on with the keys?

File: src/Type/Definition/ResolveInfo.php [hunk 2]
```diff
@@ -214,6 +216,126 @@ public function getFieldSelection(int $depth = 0): array
...
+     * For each field there is an "aliases" key regrouping all aliases of this field
+     * (if a field is not aliased, its name is present here as a key)
+     * For each of those "alias" you can find "args" key containing the arguments of the alias and the "fields" key
+     * containing the subfield of this field/alias. Each of those field have the same structure as described above.
+     *
+     * Example:
+     * query MyQuery{
+     * {
+     *   root {
+     *     id,
```
Comment (spawnia) [inline src/Type/Definition/ResolveInfo.php:231]:
```suggestion
     *     id
```

File: src/Type/Definition/ResolveInfo.php [hunk 3]
```diff
@@ -214,6 +216,126 @@ public function getFieldSelection(int $depth = 0): array
...
+    /**
+     * Helper method that returns names of all fields selected in query for
+     * $this->fieldName up to $depth levels.
+     * For each field there is an "aliases" key regrouping all aliases of this field
+     * (if a field is not aliased, its name is present here as a key)
+     * For each of those "alias" you can find "args" key containing the arguments of the alias and the "fields" key
+     * containing the subfield of this field/alias. Each of those field have the same structure as described above.
+     *
+     * Example:
+     * query MyQuery{
```
Comment (spawnia) [inline src/Type/Definition/ResolveInfo.php:228]:
```suggestion
```

File: src/Type/Definition/ResolveInfo.php [hunk 4]
```diff
@@ -214,6 +216,126 @@ public function getFieldSelection(int $depth = 0): array
...
+     *           ]
+     *       ]
+     * ]
+     *
+     * Warning: this method it is a naive implementation which does not take into account
+     * conditional typed fragments. So use it with care for fields of interface and union types.
+     * You still can alias the union type fields with the same name in order to extract their corresponding args.
+     *
+     * Example:
+     *  query MyQuery{
```
Comment (spawnia) [inline src/Type/Definition/ResolveInfo.php:295]:
```suggestion
```

File: src/Type/Definition/ResolveInfo.php [hunk 5]
```diff
@@ -214,6 +216,126 @@ public function getFieldSelection(int $depth = 0): array
...
+     *  {
+     *    root {
+     *      id,
```
Comment (spawnia) [inline src/Type/Definition/ResolveInfo.php:298]:
```suggestion
     *      id
```

File: src/Type/Definition/ResolveInfo.php [hunk 6]
```diff
@@ -266,4 +388,61 @@ private function foldSelectionSet(SelectionSetNode $selectionSet, int $descend):
...
+        foreach ($selectionSet->selections as $selectionNode) {
+            if ($selectionNode instanceof FieldNode) {
+                $fieldName = $selectionNode->name->value;
+                $aliasName = $selectionNode->alias->value ?? $fieldName;
+
+                if ($fieldName === Introspection::TYPE_NAME_FIELD_NAME) {
+                    continue;
+                }
+
+                assert($parentType instanceof HasFieldsType, 'ensured by query validation and the check above which excludes union types');
```
Comment (spawnia) [inline src/Type/Definition/ResolveInfo.php:411]: Is this copy-pasted from somewhere? The comment makes little sense to me in this context.

File: tests/Type/ResolveInfoTest.php [hunk 7]
```diff
@@ -648,4 +648,263 @@ public function testPathAndUnaliasedPathForList(): void
             ],
         ], $result);
     }
+
+    private int $aliasArgsNbTests = 0;
```
Comment (spawnia) [inline tests/Type/ResolveInfoTest.php:652]: Use a local variable that is included in the closure by reference (`use (&$var)`).

**Review by spawnia — CHANGES_REQUESTED** [review]

**Comment by spawnia** [comment]
The PHPStan error should be fixable with an `assert` call to ensure non-null.

File: src/Type/Definition/ResolveInfo.php (hunk 3 above)
Comment (zephyx) [reply src/Type/Definition/ResolveInfo.php:228]: Should I also update the one from the function getFieldSelection?

File: src/Type/Definition/ResolveInfo.php (hunk 6 above)
Comment (zephyx) [reply src/Type/Definition/ResolveInfo.php:411]: I was using the same logical as QueryPlan, we don't want stuff like __typename and we need to have fieldsType.
But I can remove this if you want.

File: src/Type/Definition/ResolveInfo.php [hunk 8]
```diff
@@ -214,6 +216,126 @@ public function getFieldSelection(int $depth = 0): array
...
+     *
+     * @throws \Exception
+     * @throws Error
+     * @throws InvariantViolation
+     *
+     * @return array<string, mixed>
+     *
+     * @api
+     */
+    public function getFieldSelectionWithAlias(int $depth = 0): array
```
[the remark this answers is not in the document]
Comment (zephyx) [reply src/Type/Definition/ResolveInfo.php:323]: 1 goes down one level, 0 stays at the root level.
This is the same logic as getFieldSelection(), for reasons of consistency.

I can go further 1 if you want, I am not expecting any side effect further 1, I have already tested this at 2 when writing the code.

-X depth = 0 depth
getFieldSelection() run like this, I don't think I should throw an error here if there are no errors in getFieldSelection().
I tried to keep the same logic.

**Comment by zephyx** [comment]
Updated

**Comment by spawnia** [comment]
I know I originally proposed it like this in https://github.com/webonyx/graphql-php/issues/1072#issuecomment-1041345447, but there we also included the type name per field. Given we no longer need that, is there any reason why we should nest aliases under a key `aliases`, as in this example:

```php
[
    'level2' => [
        'aliases' => [
            'level2000' => [
                'args' => [
                    'width' => 1,
                    'height' => 1,
                ],
            ],
            'level2' => [
                'args' => [
                    'width' => 2,
                    'height' => 20,
                ],
            ],
        ],
    ],
]
```

Or can we simplify like this?

```php
[
    'level2' => [
        'level2000' => [
            'args' => [
                'width' => 1,
                'height' => 1,
            ],
        ],
        'level2' => [
            'args' => [
                'width' => 2,
                'height' => 20,
            ],
        ],
    ],
]
```

**Comment by zephyx** [comment]
Yes, in current state, we can simplify this
But when I started writing this code, I imagined adding other information like the type you mention.
I forgot this point during my refactoring, but for example adding the type is just one line of code.
Leaving this level “aliases” would make it possible to add other information to the return if we want to evolve the function, no?

If it's ok with you, I can add this line to the code: 
$fields[$fieldName]['type'] = $fieldType->name();

This will return if you keep the types described in tests : 
```
[
    'level2' => [
        'type' => 'Int'
        'aliases' => [
            'level2000' => [
                'args' => [
                    'width' => 1,
                    'height' => 1,
                ],
            ],
            'level2' => [
                'args' => [
                    'width' => 2,
                    'height' => 20,
                ],
            ],
        ],
    ],
     'level2bis' => [
        'type' => 'level2bis'
        'aliases' => [
            'level2bis' => [
                'args' => [],
                'fields' => [
                     'level3' => [
                         'type' => 'Int',
                         'aliases' => [
                             'level3' => [
                                 'args' => [
                                     'length' => 10
                                 ]
                             ]
                         ]
                     ]
               ]
            ]
        ],
    ],
]
```

File: src/Type/Definition/ResolveInfo.php [hunk 9]
```diff
@@ -203,10 +204,135 @@ public function getFieldSelection(int $depth = 0): array
         $fields = [];
 
         foreach ($this->fieldNodes as $fieldNode) {
-            if (isset($fieldNode->selectionSet)) {
+            $selectionSet = $fieldNode->selectionSet;
```
[the remark this answers is not in the document]
Comment (spawnia) [reply src/Type/Definition/ResolveInfo.php:207]: Yup, it is just a refactoring. Given much of the code was copy-pasted, I wanted to improve the original too.

**Comment by zephyx** [comment]
> Adding the type name is kind of an arbitrary extra information anyways. Nullability or being in a list could also matter - as could other information about the field or the schema at that point.

Yes, that was my point (evolution): if latter we need more information, it would have been easy to add it alongside the "aliases" level. Once removed, we won't be able to add field related informations without introducing breaking changes.
But nevermind, I can remove it if you want.

**Comment by spawnia** [comment]
We can always add a new API if something more complex is necessary. Let's keep the simple things simple and only add complexity when it is really necessary.

**Comment by spawnia** [comment]
I renamed the nested `fields` to `selectionSet`. This terminology follows the language of the GraphQL specification when talking about queries, see https://spec.graphql.org/draft/#Field. It also clearly distinguishes this from the type level, it is about the current execution and not about what is in the schema.

**Comment by spawnia** [comment]
Thank you @zephyx, released with https://github.com/webonyx/graphql-php/releases/tag/v15.19.0.
